### PR TITLE
RavenDB-18676 - Enhance the out of memory error message

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18676

### Additional description

Enhance the out-of-memory error message

### How risky is the change?

- Low

### CODE REVIEW

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+﻿using System.Linq;
+using Sparrow.LowMemory;
+
+namespace Sparrow.Utils;
+
+public static class MemoryUtils
+{
+    public static string GetExtendedMemoryInfo(MemoryInfoResult memoryInfo)
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs:13]: You are calling this at a time where we run out of memory, but you are doing a lot of allocations.
We need to be prepared for exceptions here and return a static const string in that case.

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        var message =
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs]: Also, let's use a string builder to reduce allocations

Comment (grisha-kotler) [reply src/Sparrow/Utils/MemoryUtils.cs]: done

File: src/Sparrow/Utils/MemoryUtils.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Sparrow/Utils/MemoryUtils.cs:13]: done

File: src/Sparrow/LowMemory/LowMemoryNotification.cs [hunk 3]
```diff
@@ -43,21 +43,21 @@ internal class LowMemEventDetails
...
 
         private void RunLowMemoryHandlers(bool isLowMemory, MemoryInfoResult memoryInfo, LowMemorySeverity lowMemorySeverity = LowMemorySeverity.ExtremelyLow)
         {
             try
             {
                 try
                 {
-                    if (isLowMemory && _logger.IsOperationsEnabled)
+                    var now = DateTime.UtcNow;
+                    if (isLowMemory && _logger.IsOperationsEnabled && now - _lastLoggedLowMemory > _logLowMemoryInterval)
```
Comment (ayende) [inline src/Sparrow/LowMemory/LowMemoryNotification.cs:56]: Do we have log spam because of this?

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                }
+            }
+        }
+        catch
+        {
+            return GenericOutMemoryException;
+        }
+    }
+
+    private static string GenerateThreadsMessage()
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs]: Pass the string builder inside, why create another string?

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        try
+        {
+            var topThreadsText = new StringBuilder();
+            const int minAllocatedThresholdInBytes = 10 * 1024 * 1024;
+            var first = true;
+
+            foreach (var stats in NativeMemory.AllThreadStats
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs]: Undo the Linq here, better to avoid those methods / allocations

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                topThreadsText.Append($"name: {stats.Name}, allocations: {new Size(stats.Allocated, SizeUnit.Bytes)}");
+
+                if (stats.Count > 1)
+                {
+                    topThreadsText.Append($" (threads count: {stats.Count})");
+                }
+            }
+
+            
+            return topThreadsText.Length > 0 ? $", Top unmanaged allocations: {topThreadsText}" : null;
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs]: Allocations galore.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Sparrow/LowMemory/LowMemoryNotification.cs (hunk 3 above)
Comment (ppekrol) [reply src/Sparrow/LowMemory/LowMemoryNotification.cs:56]: we were doing that every 1s, now decreased to 5s

File: src/Sparrow/Utils/MemoryUtils.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Sparrow/Utils/MemoryUtils.cs]: done

File: src/Sparrow/Utils/MemoryUtils.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Sparrow/Utils/MemoryUtils.cs]: done

File: src/Sparrow/Utils/MemoryUtils.cs (hunk 6 above)
Comment (grisha-kotler) [reply src/Sparrow/Utils/MemoryUtils.cs]: done

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        try
+        {
+            var sb = new StringBuilder();
+            TryAppend(() => $"Commit charge: {memoryInfo.CurrentCommitCharge} / {memoryInfo.TotalCommittableMemory}, ");
+            TryAppend(() => $"Memory: {memoryInfo.TotalPhysicalMemory - memoryInfo.AvailableMemory} / {memoryInfo.TotalPhysicalMemory}, ");
+            TryAppend(() => $"Available memory for processing: {memoryInfo.AvailableMemoryForProcessing}, ");
+            TryAppend(() => $"Dirty memory: {memoryInfo.TotalScratchDirtyMemory}, ");
+            TryAppend(() => $"Managed memory: {new Size(AbstractLowMemoryMonitor.GetManagedMemoryInBytes(), SizeUnit.Bytes)}, ");
+            TryAppend(() => $"Unmanaged allocations: {new Size(AbstractLowMemoryMonitor.GetUnmanagedAllocationsInBytes(), SizeUnit.Bytes)}");
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs]: Each one of those will allocate a temporary string, no?

File: src/Sparrow/Utils/MemoryUtils.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                    {
+                        sb.Append(", Top unmanaged allocations: ");
+                        first = false;
+                    }
+                    else
+                    {
+                        sb.Append(", ");
+                    }
+
+                    TryAppend(() => $"name: {keyValue.Value}, allocations: {new Size(keyValue.Key, SizeUnit.Bytes)}");
```
Comment (ayende) [inline src/Sparrow/Utils/MemoryUtils.cs]:
```suggestion
                    TryAppend(sb => sb.Append("name: ").Append(keyValue.Value).Append(", allocations: ...
```

**Review by ayende — CHANGES_REQUESTED** [review]

**Review by ppekrol — APPROVED** [review]

File: src/Sparrow/Utils/MemoryUtils.cs (hunk 8 above)
Comment (grisha-kotler) [reply src/Sparrow/Utils/MemoryUtils.cs]: done

**Review by ayende — APPROVED** [review]

**Pull request merged**